### PR TITLE
Fixes to type subtype pairs in ege*

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

File: engine/tests/ElggCoreGetEntitiesTest.php [hunk 1]
```diff
@@ -348,6 +348,40 @@ public function testElggAPIGettersTSPValidTypeValidSubtype() {
 	}
 
 	/**
+	 * Valid type, ELGG_ENTITIES_NO_VALUE as subtype
+	 */
+	public function testElggAPIGettersTSPValidTypeNoValueSubtype() {
```
Comment (juho-jaakkola) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Needs better function name.

Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Open to suggestions.

Comment (juho-jaakkola) [inline engine/tests/ElggCoreGetEntitiesTest.php]: The existing functions also have confusing names. I suppose this is fine for now. Let's make all the names better in a separate PR.

Comment (juho-jaakkola) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Add a docblock that describes the test. That makes it easier to improve the name later.

Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Will do. I want to add another test with just an empty array.

**Comment by hypeJunction** [comment]
@juho-jaakkola I only left one test case that checks all values I could think of.

File: engine/classes/Elgg/Database/EntityTable.php [hunk 2]
```diff
@@ -821,6 +824,8 @@ function getEntityTypeSubtypeWhereSql($table, $types, $subtypes, $pairs) {
 					if ($paired_subtype_ids_str = implode(',', $paired_subtype_ids)) {
```
Comment (mrclay) [inline engine/classes/Elgg/Database/EntityTable.php]: So this case handles the NO value?

File: engine/tests/ElggCoreGetEntitiesTest.php [hunk 3]
```diff
@@ -348,6 +348,38 @@ public function testElggAPIGettersTSPValidTypeValidSubtype() {
...
+
+		foreach ($tests as $subtype) {
+			$pair = array($type => $subtype);
+			$options = array(
+				'type_subtype_pairs' => $pair,
+				'wheres' => array("(e.type != '{$type}')"),
+			);
+
+			$es = elgg_get_entities($options);
+			$this->assertFalse($es);
```
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Could you assert equal to `[]` for clarity?

File: engine/tests/ElggCoreGetEntitiesTest.php [hunk 4]
```diff
@@ -348,6 +348,38 @@ public function testElggAPIGettersTSPValidTypeValidSubtype() {
...
+			array(),
+			array(ELGG_ENTITIES_ANY_VALUE),
+			array(ELGG_ENTITIES_NO_VALUE),
+		);
+
+		foreach ($tests as $subtype) {
+			$pair = array($type => $subtype);
+			$options = array(
+				'type_subtype_pairs' => $pair,
+				'wheres' => array("(e.type != '{$type}')"),
```
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: It seems like we don't want this line. What does it do?

Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesTest.php]: We want to ensure that results do not include entities of other type.

File: engine/tests/ElggCoreGetEntitiesTest.php (hunk 3 above)
Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Why []? Ege* returns false on no match

File: engine/classes/Elgg/Database/EntityTable.php (hunk 2 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: There is too much spaghetti. It handles too many things at once.

File: engine/tests/ElggCoreGetEntitiesTest.php [hunk 5]
```diff
@@ -348,6 +348,38 @@ public function testElggAPIGettersTSPValidTypeValidSubtype() {
...
+	 * or an array with a single false/null value.
+	 * See #8851
+	 */
+	public function testElggAPIGettersTSPReturnsOnlyEntitiesOfTypeForEmptyOrFalseSubtype() {
+		$types = $this->getRandomValidTypes(1);
+		$type = $types[0];
+
+		// Test no value subtype
+		$tests = array(
+			ELGG_ENTITIES_ANY_VALUE,
```
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Don't we expect this value to return something?

File: engine/tests/ElggCoreGetEntitiesTest.php (hunk 3 above)
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: No results = empty array. Which is falsy so maybe that's why `$this->assertFalse()` works.

File: engine/tests/ElggCoreGetEntitiesTest.php (hunk 5 above)
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Setting subtype to "any value" should return something, so this test needs a closer look. I'll try to do this soon.

File: engine/tests/ElggCoreGetEntitiesTest.php (hunk 4 above)
Comment (mrclay) [inline engine/tests/ElggCoreGetEntitiesTest.php]: Then we want `=`, not `!=`. This test is making the API look a lot more competent than it is.

Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesTest.php]: The problem is that it currently returns the types it shouldn't, i.e. setting `'type_subtype_pairs' => ['groups' => []]` also returns objects and users. `!=` excludes the type we are testing and making sure we don't have other entity types in the array.

**Comment by mrclay** [comment]
Handling ELGG_ENTITIES_NO_VALUE and ELGG_ENTITIES_ANY_VALUE is still broken due to line https://github.com/Elgg/Elgg/pull/8852/files#diff-249c991a31fec2d9fe7fdb2604d226ecR789, which throws out anything falsy.
